File: leetcode/lc-us/cpp/50.cpp
class Solution {
public:
    double quick_mul(double x, long n) {
        double res = 1.0;
        while (n) {
            if (n & 1) res *= x;
            x *= x;
            n >>= 1;
        }
        return res;
    }

    double myPow(double x, int n) {
        return n > 0 ? quick_mul(x, (long) n) : 1 / quick_mul(x, -(long) n);
    }
};

File: leetcode/lc-us/rust/50.rs
fn pow(a: f64, n: i64) -> f64 {
    let mut a = a;
    let mut res = 1.0;
    let mut n = n;
    while n > 0 {
        if (n & 1) == 1 {
            res *= a;
        }
        a *= a;
        n >>= 1;
    }
    return res;
}

impl Solution {
    pub fn my_pow(x: f64, n: i32) -> f64 {
        let mut n: i64 = n as i64;
        if n < 0 {
            1.0 / pow(x, -n)
        } else {
            pow(x, n)
        }
    }
}